File: challenges/02_linked_list/python/linked_list.py
#!/usr/bin/env python3
"""LinkedList - Python 実装"""

from typing import TypeVar, Generic, Optional, Iterator, List
from dataclasses import dataclass

print("=== LinkedList Demo ===\n")

T = TypeVar("T")


# --- Node ---
@dataclass
class Node(Generic[T]):
    value: T
    next: Optional["Node[T]"] = None


# --- LinkedList ---
class LinkedList(Generic[T]):
    def __init__(self) -> None:
        self._head: Optional[Node[T]] = None
        self._tail: Optional[Node[T]] = None
        self._size: int = 0

    @property
    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    # 先頭に追加
    def push_front(self, value: T) -> "LinkedList[T]":
        node = Node(value)
        if self._head is None:
            self._head = node
            self._tail = node
        else:
            node.next = self._head
            self._head = node
        self._size += 1
        return self

    # 末尾に追加
    def push_back(self, value: T) -> "LinkedList[T]":
        node = Node(value)
        if self._tail is None:
            self._head = node
            self._tail = node
        else:
            self._tail.next = node
            self._tail = node
        self._size += 1
        return self

    # 先頭を削除
    def pop_front(self) -> Optional[T]:
        if self._head is None:
            return None
        value = self._head.value
        self._head = self._head.next
        if self._head is None:
            self._tail = None
        self._size -= 1
        return value

    # 先頭の値を参照
    @property
    def front(self) -> Optional[T]:
        return self._head.value if self._head else None

    # 末尾の値を参照
    @property
    def back(self) -> Optional[T]:
        return self._tail.value if self._tail else None

    # イテレーション
    def __iter__(self) -> Iterator[T]:
        current = self._head
        while current:
            yield current.value
            current = current.next

    def __len__(self) -> int:
        return self._size

    # 検索
    def find(self, value: T) -> Optional[Node[T]]:
        current = self._head
        while current:
            if current.value == value:
                return current
            current = current.next
        return None

    def __contains__(self, value: T) -> bool:
        return self.find(value) is not None

    # 削除
    def delete(self, value: T) -> Optional[T]:
        if self._head is None:
            return None

        # 先頭の場合
        if self._head.value == value:
            return self.pop_front()

        # それ以外
        prev = self._head
        current = self._head.next
        while current:
            if current.value == value:
                prev.next = current.next
                if current == self._tail:
                    self._tail = prev
                self._size -= 1
                return value
            prev = current
            current = current.next
        return None

    # 反転
    def reverse(self) -> "LinkedList[T]":
        if self._size <= 1:
            return self

        prev: Optional[Node[T]] = None
        current = self._head
        self._tail = self._head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        self._head = prev
        return self

    # クリア
    def clear(self) -> None:
        self._head = None
        self._tail = None
        self._size = 0

    def to_list(self) -> List[T]:
        return list(self)

    def __repr__(self) -> str:
        return f"LinkedList([{' -> '.join(map(str, self))}])"


# --- デモ ---
print("--- Basic Operations ---")

lst: LinkedList[int] = LinkedList()
print(f"Empty list: {lst}")

lst.push_front(3)
lst.push_front(2)
lst.push_front(1)
print(f"After push_front 3, 2, 1: {lst}")

lst.push_back(4)
lst.push_back(5)
print(f"After push_back 4, 5: {lst}")

print(f"Size: {lst.size}")
print(f"Front: {lst.front}")
print(f"Back: {lst.back}")

print()

print("--- Pop and Delete ---")
print(f"pop_front: {lst.pop_front()}")
print(f"After pop_front: {lst}")

print(f"delete(4): {lst.delete(4)}")
print(f"After delete(4): {lst}")

print()

print("--- Iteration ---")
lst.clear()
for x in [1, 2, 3, 4, 5]:
    lst.push_back(x)
print(f"List: {lst}")
print(f"map (x * 2): {[x * 2 for x in lst]}")
print(f"filter (x > 2): {[x for x in lst if x > 2]}")
print(f"sum: {sum(lst)}")
print(f"3 in list: {3 in lst}")

print()

print("--- Reverse ---")
print(f"Before reverse: {lst}")
lst.reverse()
print(f"After reverse: {lst}")

print()

# --- テスト ---
print("--- Tests ---")


def run_tests() -> None:
    lst: LinkedList[int] = LinkedList()

    # 空リスト
    assert lst.is_empty(), "empty? failed"
    assert lst.size == 0, "size failed"

    # push_front
    lst.push_front(1)
    lst.push_front(2)
    lst.push_front(3)
    assert lst.to_list() == [3, 2, 1], "push_front failed"

    # push_back
    lst.push_back(10)
    assert lst.to_list() == [3, 2, 1, 10], "push_back failed"

    # front / back
    assert lst.front == 3, "front failed"
    assert lst.back == 10, "back failed"

    # pop_front
    assert lst.pop_front() == 3, "pop_front failed"
    assert lst.to_list() == [2, 1, 10], "after pop_front failed"

    # delete
    assert lst.delete(1) == 1, "delete failed"
    assert lst.to_list() == [2, 10], "after delete failed"

    # reverse
    lst.push_back(20)
    lst.reverse()
    assert lst.to_list() == [20, 10, 2], "reverse failed"

    # iteration
    assert len(list(lst)) == 3, "iteration failed"
    assert 10 in lst, "contains failed"

    print("All tests passed!")


run_tests()

File: challenges/02_linked_list/ruby/linked_list.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# LinkedList - Ruby 実装

puts "=== LinkedList Demo ===\n\n"

# --- Node ---
class Node
  attr_accessor :value, :next_node

  def initialize(value)
    @value = value
    @next_node = nil
  end
end

# --- LinkedList ---
class LinkedList
  include Enumerable

  def initialize
    @head = nil
    @tail = nil
    @size = 0
  end

  attr_reader :size

  def empty?
    @size.zero?
  end

  # 先頭に追加
  def push_front(value)
    node = Node.new(value)
    if @head.nil?
      @head = node
      @tail = node
    else
      node.next_node = @head
      @head = node
    end
    @size += 1
    self
  end

  # 末尾に追加
  def push_back(value)
    node = Node.new(value)
    if @tail.nil?
      @head = node
      @tail = node
    else
      @tail.next_node = node
      @tail = node
    end
    @size += 1
    self
  end

  # 先頭を削除
  def pop_front
    return nil if @head.nil?

    value = @head.value
    @head = @head.next_node
    @tail = nil if @head.nil?
    @size -= 1
    value
  end

  # 先頭の値を参照
  def front
    @head&.value
  end

  # 末尾の値を参照
  def back
    @tail&.value
  end

  # イテレーション
  def each
    return enum_for(:each) unless block_given?

    current = @head
    while current
      yield current.value
      current = current.next_node
    end
  end

  # 検索
  def find_node(value)
    current = @head
    while current
      return current if current.value == value
      current = current.next_node
    end
    nil
  end

  # 削除
  def delete(value)
    return nil if @head.nil?

    # 先頭の場合
    if @head.value == value
      return pop_front
    end

    # それ以外
    prev = @head
    current = @head.next_node
    while current
      if current.value == value
        prev.next_node = current.next_node
        @tail = prev if current == @tail
        @size -= 1
        return value
      end
      prev = current
      current = current.next_node
    end
    nil
  end

  # 反転
  def reverse!
    return self if @size <= 1

    prev = nil
    current = @head
    @tail = @head

    while current
      next_node = current.next_node
      current.next_node = prev
      prev = current
      current = next_node
    end

    @head = prev
    self
  end

  # クリア
  def clear
    @head = nil
    @tail = nil
    @size = 0
  end

  def to_s
    "[#{to_a.join(' -> ')}]"
  end

  def inspect
    "#<LinkedList: #{self}>"
  end
end

# --- デモ ---
puts "--- Basic Operations ---"

list = LinkedList.new
puts "Empty list: #{list}"

list.push_front(3)
list.push_front(2)
list.push_front(1)
puts "After push_front 3, 2, 1: #{list}"

list.push_back(4)
list.push_back(5)
puts "After push_back 4, 5: #{list}"

puts "Size: #{list.size}"
puts "Front: #{list.front}"
puts "Back: #{list.back}"

puts

puts "--- Pop and Delete ---"
puts "pop_front: #{list.pop_front}"
puts "After pop_front: #{list}"

puts "delete(4): #{list.delete(4)}"
puts "After delete(4): #{list}"

puts

puts "--- Enumerable ---"
list.clear
[1, 2, 3, 4, 5].each { |x| list.push_back(x) }
puts "List: #{list}"
puts "map { x * 2 }: #{list.map { |x| x * 2 }}"
puts "select { x > 2 }: #{list.select { |x| x > 2 }}"
puts "reduce(:+): #{list.reduce(:+)}"

puts

puts "--- Reverse ---"
puts "Before reverse: #{list}"
list.reverse!
puts "After reverse: #{list}"

puts

# --- テスト ---
puts "--- Tests ---"

def run_tests
  list = LinkedList.new

  # 空リスト
  raise "empty? failed" unless list.empty?
  raise "size failed" unless list.size == 0

  # push_front
  list.push_front(1)
  list.push_front(2)
  list.push_front(3)
  raise "push_front failed" unless list.to_a == [3, 2, 1]

  # push_back
  list.push_back(10)
  raise "push_back failed" unless list.to_a == [3, 2, 1, 10]

  # front / back
  raise "front failed" unless list.front == 3
  raise "back failed" unless list.back == 10

  # pop_front
  raise "pop_front failed" unless list.pop_front == 3
  raise "after pop_front failed" unless list.to_a == [2, 1, 10]

  # delete
  raise "delete failed" unless list.delete(1) == 1
  raise "after delete failed" unless list.to_a == [2, 10]

  # reverse
  list.push_back(20)
  list.reverse!
  raise "reverse failed" unless list.to_a == [20, 10, 2]

  # Enumerable
  raise "each failed" unless list.map { |x| x }.size == 3
  raise "include? failed" unless list.include?(10)

  puts "All tests passed!"
end

run_tests
